internal/cephfs/core/quiesce.go

```diff
@@ -0,0 +1,237 @@
+
+	"github.com/ceph/go-ceph/cephfs/admin"
+)
+
+type QuiesceState string
+
+const (
+	Released  QuiesceState = "RELEASED"
+	Quiescing QuiesceState = "QUIESCING"
+	Quiesced  QuiesceState = "QUIESCED"
+)
+
+// GetQuiesceState returns the quiesce state of the filesystem.
+func GetQuiesceState(set admin.QuiesceState) QuiesceState {
+	var state QuiesceState
+	switch set.Name {
+	case "RELEASED":
+		state = Released
+	case "QUIESCING":
+		state = Quiescing
+	case "QUIESCED":
+		state = Quiesced
+	default:
+		state = QuiesceState(set.Name)
+	}
+
+	return state
+}
+
+type FSQuiesceClient interface {
+	// FSQuiesce quiesces the subvolumes in the filesystem.
+	FSQuiesce(
+		ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// GetVolumes returns the list of volumes in the filesystem that are to be
+	// quiesced.
+	GetVolumes() []Volume
+	// FSQuiesceWithExpireTimeout quiesces the subvolumes in the
+	// filesystem with an expiration timeout.
+	FSQuiesceWithExpireTimeout(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// ResetFSQuiesce resets the quiesce timeout for the subvolumes in
+	// the filesystem.
+	ResetFSQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// ReleaseFSQuiesce releases the quiesce on the subvolumes in the
+	// filesystem.
+	ReleaseFSQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+}
+
+type Volume struct {
+	VolumeID  string
+	ClusterID string
+}
+
+type fsQuiesce struct {
+	fsName  string
+	volumes []Volume
+	// subVolumeGroupMapping is a map of subvolumes to groups.
+	subVolumeGroupMapping map[string][]string
+	fsa                   *admin.FSAdmin
+}
+
+// NewFSQuiesce returns a new instance of fsQuiesce. It
+// take the filesystem name, the list of volumes to be quiesced, the mapping of
+// subvolumes to groups and the cluster connection as input.
+func NewFSQuiesce(
+	fsName string,
+	volumes []Volume,
+	mapping map[string][]string,
+	conn *util.ClusterConnection,
+) (FSQuiesceClient, error) {
+	fsa, err := conn.GetFSAdmin()
+	if err != nil {
+		return nil, err
+	}
+
+	return &fsQuiesce{
+		fsName:                fsName,
+		volumes:               volumes,
+		subVolumeGroupMapping: mapping,
+		fsa:                   fsa,
+	}, nil
+}
+
+// GetVolumes returns the list of volumes in the filesystem that are to be
+// quiesced.
+func (fq *fsQuiesce) GetVolumes() []Volume {
+	return fq.volumes
+}
+
+// getMembers returns the list of names in the format
+// group/subvolume that are to be quiesced. This is the format that the
+// ceph fs quiesce expects.
+// Example: ["group1/subvolume1", "group1/subvolume2", "group2/subvolume1"].
+func (fq *fsQuiesce) getMembers() []string {
+	volName := []string{}
+	for svg, sb := range fq.subVolumeGroupMapping {
+		for _, s := range sb {
+			name := svg + "/" + s
+			volName = append(volName, name)
+		}
+	}
+
+	return volName
+}
+
+func (fq *fsQuiesce) FSQuiesce(
```

Comment:
What is the difference with `FSQuiesceWithExpireTimeout()`? If both functions are needed, please comment them so that it is clear when to use one of them, and not the other.

Reply:
added a comment when to use `FSQuiesceWithExpireTimeout`

Reply:
Thanks, but I still do not see a difference in the functions. Both seem to pass the same Timeout + Expiration?

Reply:
Yes for now it looks the same, based on the testing we might change the timeout and expiration for both, for easy reading/maintenance i have kept 2 separate functions for now.

Reply:
If this is a TODO, please add that as a comment, so that grepping for TODO in the sources finds it again.

Not sure if it is beneficial to keep two identical functions. Please consider removing the duplicate and commenting TODO in the code where a (non)timeout function may be preferable.